api/wasm/v0/cpp/proxy_wasm_externs.h

```diff
@@ -0,0 +1,121 @@
+ * @param root_context_id is a unique identifier for the configuration verification context.
+ * @param configuration_size is the size of any configuration available via
+ * proxy_get_configuration().
+ * @return non-zero on success and zero on failure (i.e. bad configuration).
+ */
+extern "C" uint32_t proxy_validate_configuration(uint32_t root_context_id,
+                                                 uint32_t configuration_size);
+/**
+ * Called when a plugin loads or when plugin configuration changes dynamically.
+ * @param root_context_id is an identifier for one or more related plugins.
+ * @param configuration_size is the size of any configuration available via
+ * proxy_get_configuration().
+ * @return non-zero on success and zero on failure (e.g. bad configuration).
+ */
+extern "C" uint32_t proxy_on_configure(uint32_t root_context_id, uint32_t configuration_size);
+
+// Stream calls.
+
+/**
+ * Called when a request, stream or other ephemeral context is created.
+ * @param context_id is an identifier the ephemeral context.
+ * @param configuration_size is the size of any configuration available via
+ * proxy_get_configuration().
+ */
+extern "C" void proxy_on_create(uint32_t context_id, uint32_t root_context_id);
+
+// Stream and Non-stream calls.
+
+/**
+ * For stream contexts, called when the stream has completed. Note: if applicable proxy_on_log() is
+ * called after proxy_on_done() and before proxy_on_delete(). For root contexts, proxy_on_done() is
+ * called when the VM is going to shutdown.
+ * @param context_id is an identifier the context.
+ * @return non-zero to indicate that this context is done. Stream contexts must return non-zero.
+ * Root contexts may return zero to defer the VM shutdown and the proxy_on_delete call until after a
+ * future proxy_done() call by the root context.
+ */
+extern "C" uint32_t proxy_on_done(uint32_t context_id);
+/**
+ * Called when the context is being deleted and will no longer receive any more calls.
+ * @param context_id is an identifier the context.
+ */
+extern "C" void proxy_on_delete(uint32_t context_id);
+
+//
+// ABI calls from the VM to the host.
+//
+
+// Configuration and Status
+
+/**
+ * Called from the VM to get any configuration. Valid only when in a proxy_on_start(),
+ * proxy_validate_configuration() or proxy_on_configure() handler.
+ * @param configuration_ptr a pointer to a location which will be filled with either nullptr (if no
+ * configuration is available) or a pointer to a malloc()ed block containing the configuration
+ * bytes.
+ * @param configuration_size a pointer to a location containing the size (or zero) of any returned
+ * configuration byte block.
+ * @return a WasmResult: OK, InvalidMemoryAccess. Note: if OK is returned  *configuration_ptr may
+ * be nullptr.
+ */
+extern "C" WasmResult proxy_get_configuration(const char** configuration_ptr,
+                                              size_t* configuration_size);
```

Comment:
Stupid question: why do we do this this way? Why don't we just pass the config in with the other call from host -> vm?

Reply:
Because this requires allocating memory in the VM.  The callee is now required to free that memory otherwise we get a memory leak.  This caused problems for early plugin users.  For the ABI @PiotrSikora and I decided to be consistent and always require that any object which would be owned by the code in the VM would be requested explicitly by the code in the VM. We also decoded that all ABI calls would return an element of WasmResult.   The language specific APIs are much higher level and can use GC/smart pointer and other language specific mechanisms to make developing plugins convenient, but for the ABI we felt that consistency was valuable.

Reply:
OK far enough. Please cover this in the high level design doc that has been requested. Thanks.

Reply:
Will do!